hilti/toolchain/include/ast/node-collection.h

```diff
@@ -0,0 +1,334 @@
+    SetIterator() {}
+    ~SetIterator() = default;
+
+    const Node& node() const { return *_iter; }
+
+    SetIterator& operator=(const SetIterator& other) = default;
+    SetIterator& operator=(SetIterator&& other) = default;
+    const T& operator*() const { return value(); }
+    const T* operator->() const { return &value(); }
+    bool operator==(const SetIterator& other) const { return _iter == other._iter; }
+    bool operator!=(const SetIterator& other) const { return ! (*this == other); }
+
+    SetIterator operator++(int) {
+        auto x = SetIterator(_iter);
+        ++_iter;
+        return x;
+    }
+
+    SetIterator& operator++() {
+        ++_iter;
+        return *this;
+    }
+
+    SetIterator& operator+=(difference_type i) {
+        _iter += i;
+        return *this;
+    }
+
+    SetIterator& operator-=(difference_type i) {
+        _iter -= i;
+        return *this;
+    }
+
+    difference_type operator-(const SetIterator& other) const { return _iter - other._iter; }
+    SetIterator operator-(difference_type i) const { return SetIterator(_iter - i); }
+    SetIterator operator+(difference_type i) const { return SetIterator(_iter + i); }
+
+private:
+    const T& value() const { return ((*_iter).get()); }
+
+    BaseIterator _iter;
+};
+
+/**
+ * A set of AST nodes. The set creates a view of nodes that can be iterated
+ * over, yielding a reference to each node in turn. In contrast to `Range`, a
+ * set can include nodes that are not all part of a continous slice inside a
+ * vector.
+ */
+template<typename T>
+class Set {
+public:
+    using iterator = SetIterator<T>;
+    using const_iterator = SetIterator<T>;
+
+    Set() {}
+    Set(const Set& other) = default;
+    Set(Set&& other) = default;
+    ~Set() = default;
+
+    auto begin() const { return const_iterator(_set.begin()); }
+    auto end() const { return const_iterator(_set.end()); }
+    auto size() const { return _set.size(); }
+    bool empty() const { return _set.empty(); }
+    void insert(const T& t) { _set.push_back(t); }
+
+    /**
+     * Returns a new vector containing copies of all nodes that the range
+     * includes.
+     **/
+    std::vector<T> copy() const {
+        std::vector<T> x;
+        for ( auto i = begin(); i != end(); i++ )
+            x.push_back(*i);
+
+        return x;
+    }
+
+    const T& operator[](size_t i) const { return *(begin() + i); }
+
+    bool operator==(const Set& other) const {
+        if ( this == &other )
+            return true;
+
+        if ( size() != other.size() )
+            return false;
+
+        auto x = begin();
+        auto y = other.begin();
+        while ( x != end() ) {
+            if ( ! (*x++ == *y++) )
+                return false;
+        }
+
+        return true;
+    }
+
+    Set& operator=(const Set& other) = default;
+    Set& operator=(Set&& other) = default;
+
+private:
+    std::vector<std::reference_wrapper<const T>> _set;
+};
+
+/**
```

Comment:
Comment for all these helper functions: I'd much prefer if we'd make the existing functions in `util` generic enough that they can deal with arbitrary containers. This should probably be pretty straight-forward with return type computed with `declval` (and for `transform` additionally `result_of`). Insertion can be performed generically with `back_inserter`.

Reply:
I'll create a ticket for this.